pysrc/bytewax/recovery.py

```diff
@@ -1,15 +1,202 @@
+
+```
+$ python -m bytewax.recovery db_dir/ 4
+```
+
+The second parameter (e.g. `4`) is the number of recovery partitions
+to create. This number is fixed and cannot be changed later without
+manual SQLite interventions. In general, we recommend picking a number
+of partitions equal to the maximum number of workers you think you
+will ever rescale to.
+
+This will create a set of partitions:
+
+```
+$ ls db_dir/
+part-0.sqlite3
+part-1.sqlite3
+part-2.sqlite3
+part-3.sqlite3
+```
+
+Once the recovery partition files have been created, they must be
+placed in locations that are accessible to the workers. The cluster
+has a whole must have access to all partitions, but any given worker
+need not have access to any partition in particular (or any at
+all). It is ok if a given partition is accesible by multiple workers;
+only one worker will use it.
+
+Although the partition init script will not result in these,
+partitions after execution may consist of multiple files:
+
+```
+$ ls db_dir/
+part-0.sqlite3
+part-0.sqlite3-shm
+part-0.sqlite3-wal
+part-1.sqlite3
+part-2.sqlite3
+part-3.sqlite3
+part-3.sqlite3-shm
+part-3.sqlite3-wal
+```
+
+You must move the files with the prefix `part-*.` all together.
+
+In general, you'll want to your environment to try and evenly
+distribute partitions among your workers.
+
+If you are not running in a cluster environment but on a single
+machine, placing all the partitions in a single local filesystem
+directory is fine.
+
+Execution
+---------
+
+To enable recovery when you execute a dataflow, pass the `-r` flag to
+`bytewax.run` and specify the recovery directory.
+
+```
+$ python -m bytewax.run ... -r db_dir/
+```
+
+See the module docstring for `bytewax.run` for more information.
+
+Resume
+------
+
+If a dataflow aborts, abruptly shuts down, or gracefully exits due to
+EOF, you can resume the dataflow via running it again pointing at the
+same recovery directory. Bytewax will automatically find the most
+recent consistent snapshot to resume from.
+
+This requires that the next execution has access to all of the
+recovery partitions.
+
+This next execution does not need to have the same number of workers
+as the previous one; you are allowed to rescale the cluster and state
+will find its way to the proper workers.
+
+If you want to fully restart a dataflow at the beginning of input and
+ignore all previous state, delete partitions in the recovery
+directory.
+
+Epochs
+------
+
+The **epoch interval** is the system time interval at which an
+execution cluster synchronizes and snapshots its progress and
+state. You can adjust this duration via the `-e` parameter to
+`bytewax.run`. It defaults to every 10 seconds.
+
+The dataflow can only resume on epoch boundaries.
+
+In general, the longer this duration is, the less overhead there will
+be while the dataflow is executing due to recovery data, but the
+further back the dataflow might have to resume from.
+
```

Comment:
Is it worth having a discussion about naming around Epochs here? I feel like epochs are largely opaque to the user and the most meaningful piece of information is around recovery.

recovery-interval
recovery-epoch

Reply:
Like it!